src/io_callbacks.zig

```diff
@@ -0,0 +1,463 @@
+const std = @import("std");
+const assert = std.debug.assert;
+const os = std.os;
+const linux = os.linux;
+const IO_Uring = linux.IO_Uring;
+const io_uring_cqe = linux.io_uring_cqe;
+const io_uring_sqe = linux.io_uring_sqe;
+
+pub const IO = struct {
+    /// An intrusive first in/first out linked list.
+    const Fifo = struct {
+        in: ?*Completion = null,
+        out: ?*Completion = null,
+
+        fn push(fifo: *Fifo, completion: *Completion) void {
+            assert(completion.next == null);
+            if (fifo.in) |in| {
+                in.next = completion;
+                fifo.in = completion;
+            } else {
+                assert(fifo.out == null);
+                fifo.in = completion;
+                fifo.out = completion;
+            }
+        }
+    };
+
+    ring: IO_Uring,
+
+    /// Completions not yet submitted to the kernel and waiting on available space in the
```

Comment:
Operations instead of Completions? A completion is more like a CQE, and this is not an SQE either because we're not yet in the queue.